### PR TITLE
Feat/ansible inventory

### PR DESCRIPTION
**Pull request by vincentc-afk**

### CODE REVIEW

File: Source/Ansible/ansible-inventory.env [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+###### DOCUMENTATION BLOC #####
+########### CODE BLOC #########
+    local _t4dErrorCode=1
+    local _File="$1"
+    local _Group="$2"
+
+    local _groups=$(grep -e '^\[' $_File)
+
+    if [[ "$(echo "$_groups" | grep -w "$_Group")" == "" ]]; then
+        echo "false"
```
Comment (QuentinNIO) [inline Source/Ansible/ansible-inventory.env:68]: we talk about having a better way to output this

**Review by QuentinNIO — CHANGES_REQUESTED** [review]

File: Source/Ansible/_main_.env [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    local _Action="$2"
+    local _File="$1"
+    local _Group="$3"
+    local _Host="$4"
+
+    _t4dCheckFile $_File
+
+    case "$_Action" in
+        add|a)
+            _t4dCheckVariable $_Group $_Host
```
Comment (QuentinNIO) [inline Source/Ansible/_main_.env]: remove $ from args and update _CheckVariable behavior

File: Source/Ansible/ansible-inventory.env [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+#       {XXXXXXX}       = Value To Be Changed, if {URL} then replace {URL} with a valid URL
+###
+###### DOCUMENTATION BLOC #####
+########### CODE BLOC #########
+    local _t4dErrorCode=1
+    local _File="$1"
+    local _Group="$2"
+    local _Host="$3"
+
+    _t4dSrcYamlOpenFile "$_File"
```
Comment (QuentinNIO) [inline Source/Ansible/ansible-inventory.env:26]: did we not create a specific function for openning inventory files ?

File: Source/Ansible/ansible-inventory.env [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+# Generated From $Tools4Dev_PATH/Engine/G4d/templates.env
+# v1.8.0
+###### _t4dSrcAnsibleInventoryFind
+#   - Name:
+#       _t4dSrcAnsibleInventoryFind
+###
+#   - Synopsis:
+#       find in inventory
+###
+#   - Usage:
```
Comment (QuentinNIO) [inline Source/Ansible/ansible-inventory.env:48]: we might want to have a way to send an already oppened file in this function

File: Source/Ansible/ansible-inventory.env [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+#        _t4dSrcAnsibleInventoryFind <file>
+###
+#   - Note:
+#       <Required>
+#       [Optionnal]
+#       {XXXXXXX}       = Value To Be Changed, if {URL} then replace {URL} with a valid URL
+###
+###### DOCUMENTATION BLOC #####
+########### CODE BLOC #########
+    cat "$1" | cut -d '|' -f1 | rev | grep -Eo '>.*' | rev | sort | uniq | sed 's|>$||g'
```
Comment (QuentinNIO) [inline Source/Ansible/ansible-inventory.env:95]: you did not use `chf` to create this. 

Please add 
`local _t4dErrorCode=1`
and `local _file=$1`

and do not cat if file does not exist. 

`if [[ -e $_file ]] `

File: Source/Yaml/_main_.env [hunk 6]
```diff
@@ -171,8 +171,33 @@ _t4dSrcYamlUpdateLine(){
...
         else
-            _t4dDebugLog $perror "Path not found >$_Path<"
-            return $_t4dErrorCode
+            _t4dDebugLog $pinfo "Creating $_Path"
+            local _Name="$(basename "$(echo $_Path | sed 's|\.|/|g' | sed 's|>|/|g')")"
+            local _Type="$(echo "${_Path}" | grep -Eo ".$_Name$" | cut -c-1 )"
+            local _RootPath="$(echo "${_Path}" | sed "s|.${_Name}$||g")"
+            
+
+            local _ParentLine="$(cat $_File | sed "s| |§SPACE§|g" | grep "^$_RootPath|")"
```
Comment (QuentinNIO) [inline Source/Yaml/_main_.env:180]: you forgot to add the statement that we do not handle creating a path if rootPath does not exist and is not lvl0 dept

File: Source/Yaml/_main_.env [hunk 7]
```diff
@@ -181,3 +206,40 @@ _t4dSrcYamlUpdateLine(){
...
+#       [Optionnal]
+#       {XXXXXXX}       = Value To Be Changed, if {URL} then replace {URL} with a valid URL
+###
+###### DOCUMENTATION BLOC #####
+########### CODE BLOC #########
+    local _t4dErrorCode=1
+    local _File="$1"
+    local _YmlArg="$2"
+
+    cat $_File | grep "$_YmlArg" > "$_File.tmp"
```
Comment (QuentinNIO) [inline Source/Yaml/_main_.env:234]: you should never use argument without adding quotes, it's a really bad way to do shell, even more with cat

File: Source/Yaml/yaml-parser.env [hunk 8]
```diff
@@ -138,3 +138,45 @@ _t4dSrcYamlParseFileRevert(){
...
+###
+###### DOCUMENTATION BLOC #####
+########### CODE BLOC #########
+    local _t4dErrorCode=1
+    local _File="$1"
+    local _InPutParsedFile="$2"
+
+    _t4dSrcYamlParseClean $_InPutParsedFile
+
+    local _Content="$(cat $_File | sed "s| |§SPACE§|g")"
```
Comment (QuentinNIO) [inline Source/Yaml/yaml-parser.env:168]: also here

**Review by QuentinNIO — CHANGES_REQUESTED** [review]

File: Source/Yaml/_main_.env (hunk 6 above)
Comment (vincentc-afk) [reply Source/Yaml/_main_.env:180]: I will call you afterwards for this point

File: Source/Ansible/ansible-inventory.env [hunk 9]
```diff
@@ -92,7 +92,15 @@ _t4dSrcAnsibleSortGroups(){
...
 ###### DOCUMENTATION BLOC #####
 ########### CODE BLOC #########
-    cat "$1" | cut -d '|' -f1 | rev | grep -Eo '>.*' | rev | sort | uniq | sed 's|>$||g'
+    local _t4dErrorCode=1
+    local _File=$1
+
+    if [[ -e $_file ]];then
+        cat "$_File" | cut -d '|' -f1 | rev | grep -Eo '>.*' | rev | sort | uniq | sed 's|>$||g'
+    else
+        _t4dDebugLog $perror "empty file"
```
Comment (QuentinNIO) [inline Source/Ansible/ansible-inventory.env:101]: Update log -> File >$_File< Does not exist. 

You need to be sure that user (or even you in 3 months) will be able to understand what's going wrong

File: Source/Yaml/yaml-parser.env [hunk 10]
```diff
@@ -165,13 +165,15 @@ _t4dSrcYamlParseInventory(){
...
-    local _Content="$(cat $_File | sed "s| |§SPACE§|g")"
+    local _Content="$(cat "$_File" | sed "s| |§SPACE§|g")"
     local _GroupName=""
 
     for _line in $(echo $_Content); do
         if [[ "$(echo $_line | grep '^\[')" != "" ]]; then
             _GroupName="$(echo $_line | sed 's|\[|§DQUOTE§|g' | sed 's|\]|§DQUOTE§|g' )"
             echo ">groups~${_GroupName}~|${_GroupName}|0|LIST|groups|§DASH§" >> $_InPutParsedFile
+        elif [[ "$(echo $_line | grep '^\#')" != "" ]];then
+            echo "|§SPACE§§SPACE§${_line}§SPACE§§SPACE§|0|COMMENT||"
```
Comment (QuentinNIO) [inline Source/Yaml/yaml-parser.env:176]: why are you adding spaces ?

**Review by QuentinNIO — CHANGES_REQUESTED** [review]

**Review by QuentinNIO — APPROVED** [review]

**Pull request merged**